### ISSUE TITLE
shape mismatch when i use model optimizer for  Deconvolution layer, the "output_padding" parameter doesn't seem to be working

### ISSUE THREAD

**Issue by Sanyuanliu** [openvinotoolkit/openvino#202]
### Here isthe debug info for  model optimizer:
Op: Deconvolution
[ 2019-07-16 21:51:00,931 ] [ DEBUG ] [ infer:140 ]  Inputs:
[ 2019-07-16 21:51:00,931 ] [ DEBUG ] [ infer:34 ]  input[0]: shape = [  1 128  16  16], value = <UNKNOWN>
[ 2019-07-16 21:51:00,931 ] [ DEBUG ] [ infer:34 ]  input[1]: shape = [128 128   3   3], value = [[[[-4.5972541e-02 -1.9228101e-02  6.3651167e-02]
   [-5.5916518e-02  5.1128842e-02 -5.4433964e-0...
[ 2019-07-16 21:51:00,931 ] [ DEBUG ] [ infer:34 ]  input[2]: shape = [128], value = [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0....
[ 2019-07-16 21:51:00,931 ] [ DEBUG ] [ infer:142 ]  Outputs:
[ 2019-07-16 21:51:00,931 ] [ DEBUG ] [ infer:34 ]  output[0]: shape = [  1 128  31  31], value = <UNKNOWN>

### The parameter setting for the Deconvolution op:
nn.Conv2DTranspose(channels=out_channels, kernel_size=3, padding=1,strides=2,output_padding=1)

### The shape calculate method give by mxnet document in https://beta.mxnet.io/api/gluon/_autogen/mxnet.gluon.nn.Conv2DTranspose.html

out_height = (height-1)*strides[0]-2*padding[0]+kernel_size[0]+output_padding[0]
out_width = (width-1)*strides[1]-2*padding[1]+kernel_size[1]+output_padding[1]

The input shape is 16* 16, the output shape should be 32 x 32,  but the model optimizer calculation result is 31 x 31, the "output_padding" parameter doesn't seem to be working!

**Comment by shubha-ramani** [comment]
Dear Sanyuanliu,
So per your discussion above 

> nn.Conv2DTranspose(channels=out_channels, kernel_size=3, padding=1,strides=2,output_padding=1)

If a given input layer has the above dimensions,
And using this formula output_Size = ((input_size + 2*padding - filter_size)/stride) +1 we have output_size =  (16 + 2*1 + 3)/2 = 21/2 = 10.5 ~ 11 for the output size.

Am I missing something or can you clarify ? Why do you think the output shape should be 32*32 ?

Thanks,

Shubha

**Comment by Sanyuanliu** [comment]
Dear shubha-ramani,
Thank you for your reply, the formula you have given is for Convolution op,  since the op is Deconvolution, the output size should use this formula:
out_height = (height-1)strides[0]-2padding[0]+kernel_size[0]+output_padding[0], which is output_size =(16-1)*2-2+3+1=32.
Thanks

**Comment by shubha-ramani** [comment]
Dear @Sanyuanliu Oops (stupid me) you are right. I missed that it was actually DeConvolution. Please give me the exact model optimizer command you used. Also can you attach a main.cpp with your code ? Please don't give me a huge project. Just a main.cpp which looks like a sample. It looks like a bug to me. Sorry for the inconvenience !

Shubha

**Comment by Sanyuanliu** [comment]
Hello Shubha,

thanks for your reply,  here are my model optimizer command：

`python mo_mxnet.py --input_model E:\DeconvNet-0000.params --input_shape [1,3,512,512] --output_dir E:\ --log_level DEBUG`

The DeconvNet-0000.params was generated by this python script:
[DeconvNet.py.txt](https://github.com/opencv/dldt/files/3409094/DeconvNet.py.txt)



Sanyuanliu

**Comment by shubha-ramani** [comment]
Dear @Sanyuanliu 
I'm not even getting the output you are getting, but of course I'm using R2 (should be updated to dldt soon). Nevertheless, I get an error which you don't mention, and I got it on 2019R2 and 2019R1.1 . I used your [DeconvNet.py](https://github.com/opencv/dldt/files/3409094/DeconvNet.py.txt) to generate the *.json and params file. And I used your above model optimizer command. Except I get the following error:

> [ ERROR ]  Shape [  1 256  -1  -1] is not fully defined for output 0 of "linknet_v10__plus0". Use --input_shape with positive integers to override model input shapes.
[ ERROR ]  Cannot infer shapes or values for node "linknet_v10__plus0".
[ ERROR ]  Not all output shapes were inferred or fully defined for node "linknet_v10__plus0".
 For more information please refer to Model Optimizer FAQ (https://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_Model_Optimizer_FAQ.html), question #40.
[ ERROR ]
[ ERROR ]  It can happen due to bug in custom shape infer function <function Elementwise.__init__.<locals>.<lambda> at 0x0000016B65262B70>.

The above error doesn't make sense to me actually. As aforementioned I get it on 2019R1 as well as 2019R2. 

I noticed that your model uses as an inner core   **vision.resnet18_v1()**

Note that from our [Model Optimizer MxNet Docs](http://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_MxNet.html) ResNet-152 v1 is a supported topology while resnet18_v1() is not mentioned. 

Model Optimizer's error seems like a Model Optimizer bug but then again, it never advertised to support resnet18_v1().  My hope is that if you use a supported MxNet model as your inner core, Model Optimizer should not throw an error. And if it still throws an error in this case, then it's a bug.

Please reply back with your thoughts.

Thanks,

Shubha

**Comment by Sanyuanliu** [comment]
Dear @shubha-ramani 
Sorry for replying so late, when i used the model optimizer command, i got the same error information as you mentioned. So i set "--log_level DEBUG",  I get the following information:
'''
 [ DEBUG ] [ infer:129 ]  Op: elemwise_add
 [ DEBUG ] [ infer:140 ]  Inputs:
 [ DEBUG ] [ infer:34 ]  input[0]: shape = [  1 256  31  31], value = < UNKNOWN>
 [ DEBUG ] [ infer:34 ]  input[1]: shape = [  1 256  32  32], value = < UNKNOWN>
[ DEBUG ] [ infer:142 ]  Outputs:
[ DEBUG ] [ infer:34 ]  output[0]: shape = [  1 256  -1  -1], value = < UNKNOWN>
[ ERROR ]  Shape [  1 256  -1  -1] is not fully defined for output 0 of "linknet_v10__plus0". Use --input_shape with positive integers to override model input shapes.
[ ERROR ]  Cannot infer shapes or values for node "linknet_v10__plus0".
[ ERROR ]  Not all output shapes were inferred or fully defined for node "linknet_v10__plus0".
'''
The error was caused by the shape mismatch  for Deconvolution layer, the "output_padding" parameter doesn't seem to be working as i mentioned  from the beginning.

Besides, i  made a sample test with a single Deconvolution layer in the Network 
[DeconvLayer.py.txt](https://github.com/opencv/dldt/files/3479598/DeconvLayer.py.txt) 
I got the following debug information:
'''
 [ DEBUG ] [ infer:129 ]  Op: Deconvolution
 [ DEBUG ] [ infer:140 ]  Inputs:
[ DEBUG ] [ infer:34 ]  input[0]: shape = [  1   1 512 512], value = < UNKNOWN>
[ DEBUG ] [ infer:34 ]  input[1]: shape = [1 1 3 3], value = [[[[ 0.04819721  0.01438687  0.05011239]
   [ 0.00628365  0.04861524 -0.01068833]   [ 0.01729892...
 [ DEBUG ] [ infer:34 ]  input[2]: shape = [1], value = [0.]
 [ DEBUG ] [ infer:142 ]  Outputs:
 [ DEBUG ] [ infer:34 ]  output[0]: shape = [   1    1 1023 1023], value = < UNKNOWN>
 [ DEBUG ] [ infer:127 ]  --------------------
 [ DEBUG ] [ infer:128 ]  Partial infer for upsampleblock0_deconvbnrelu0_batchnorm0_fwd
'''

The output shape(1023) still mismatch with the calculate result from the mxnet document :((512-1)*2-2*1+3+1=1024).

Thanks,
SanyuanLiu

**Comment by shubha-ramani** [comment]
Dear @Sanyuanliu 

Thanks for the debug.  Allow me to reproduce this issue with your revised DeconvLayer.py . If I can reproduce it, I will file a bug !

thanks,

Shubha

**Comment by shubha-ramani** [comment]
Dear @Sanyuanliu ,
This time I reproduced your bug. Yes I see dim = 1023 in the IR but according to the MXNet formula (in the documentation you refer) it should be 1024.

Thanks for your patience !

Shubha

**Comment by lazarevevgeny** [comment]
Verified that the model is converted correctly (output spatial shape is 32*32) with the Model Optimizer from the master branch.
